src/test/kotlin/com/ampnet/auditornode/script/api/classes/WebSocketInputUnitTest.kt

```diff
@@ -0,0 +1,286 @@
+                on { hasMember(DESCRIPTION) } doReturn true
+                on { getMember(DESCRIPTION) } doReturn description2Value
+            }
+
+            val stringTypeValue = mock<Value> {
+                on { isString } doReturn true
+                on { asString() } doReturn InputType.STRING.name
+            }
+            val name3Value = mock<Value> {
+                on { isString } doReturn true
+                on { asString() } doReturn "name3"
+            }
+            val description3Value = mock<Value> {
+                on { isString } doReturn true
+                on { asString() } doReturn "description3"
+            }
+            val stringField = mock<Value> {
+                on { hasMember(TYPE) } doReturn true
+                on { getMember(TYPE) } doReturn stringTypeValue
+                on { hasMember(NAME) } doReturn true
+                on { getMember(NAME) } doReturn name3Value
+                on { hasMember(DESCRIPTION) } doReturn true
+                on { getMember(DESCRIPTION) } doReturn description3Value
+            }
+
+            val arrayMock = mock<Value> {
+                on { hasArrayElements() } doReturn true
+                on { arraySize } doReturn 9
+                on { getArrayElement(0) } doReturn fieldWithNoType
+                on { getArrayElement(1) } doReturn fieldNonStringType
+                on { getArrayElement(2) } doReturn fieldWithUnknownType
+                on { getArrayElement(3) } doReturn fieldWithNoName
+                on { getArrayElement(4) } doReturn fieldWithNoDescription
+                on { getArrayElement(5) } doReturn booleanField
+                on { getArrayElement(6) } doReturn numberField
+                on { getArrayElement(7) } doReturn stringField
+                on { getArrayElement(8) } doReturn null
+            }
+
+            val message = "test message"
+            val result = service.readFields(arrayMock, message)
+
+            assertThat(result)
+                .isNotNull()
+                .isEqualTo(
+                    MapApi(
+                        mapOf(
+                            "name1" to booleanValue,
+                            "name2" to numberValue,
+                            "name3" to stringValue
+                        )
+                    )
+                )
+            then(webSocketApi)
+                .should(times(1))
+                .sendCommand(
+                    ReadFieldsCommand(
+                        message,
+                        listOf(
+                            InputField(InputType.BOOLEAN, "name1", "description1"),
+                            InputField(InputType.NUMBER, "name2", "description2"),
+                            InputField(InputType.STRING, "name3", "description3")
+                        )
+                    )
+                )
+        }
+    }
+
+    @Test
+    fun `must correctly send web socket command for button() call and take value from queue`() {
```

Comment:
Call to `service.readFields(arrayMock, message)` does that. If we want to check the queue size directly, we would need to make it non-private, which I don't really like... In any case, it is not any better that calling `readFields` since it uses `queue.take()` which would block if there were no values available. Documentation of `queue.take()` should be a guarantee enough that item was taken from the queue - we are not testing the Java standard library here.

Reply:
Hmm I don't really see a way to do that here without making `queue` non-private in `WebSocketInput` or adding a method to access the queue which won't be used in the production code anyway. In my opinion, this is not to be checked in unit test anyway - that's why we have integration and API tests which can check for such a thing. In the unit test we only check if the `button` call sends a web socket message and returns immediately when there is something in the queue.

Reply:
I understand it, this was just a thought.

Reply:
Actually, I got an idea yesterday evening. I will implement it.